### PR TITLE
Add carbon simulation and KPP mechanism

### PR DESCRIPTION
**Pull request by msulprizio**
This PR adds a carbon (CO-CO2-CH4-OCS) simulation and KPP mechanism addressing feature request issue https://github.com/geoschem/geos-chem/issues/654.  It supersedes PR https://github.com/geoschem/geos-chem/pull/1251. 

TO DO:

- [x] Rename `carboncycle` to `carbon`
- [ ] Scientific validation for:
  - [ ] CH4 (@msulprizio)
  - [ ] CO (@jennyfisher @BeaBu)
  - [ ] CO2 (@kbowman77)
  - [ ] OCS (@kbowman77)

### CODE REVIEW

**Comment by jennyfisher** [comment]
Can we please add @BeaBu (Beata Bukosa) for the scientific validation? Thanks!

**Comment by yantosca** [comment]
@jennyfisher: I have tagged @BeaBu on this PR.  Thanks!

**Comment by msulprizio** [comment]
These updates are now passing integration tests for both GCClassic and GCHP (see below). Note: GCHP integration tests for the carbon cycle simulation will be added with PR https://github.com/geoschem/geos-chem/pull/1565.

```
==============================================================================
GEOS-Chem Classic: Execution Test Results

Using 24 OpenMP threads
Number of execution tests: 29
==============================================================================
 
Execution tests:
------------------------------------------------------------------------------
gc_025x03125_CH4_geosfp_47L_na......................Execute Simulation....PASS
gc_025x03125_fullchem_geosfp_47L_na.................Execute Simulation....PASS
gc_05x0625_CH4_merra2_47L_na........................Execute Simulation....PASS
gc_05x0625_fullchem_merra2_47L_na...................Execute Simulation....PASS
gc_4x5_aerosol_merra2...............................Execute Simulation....PASS
gc_4x5_carbon_merra2................................Execute Simulation....PASS
gc_4x5_CH4_merra2...................................Execute Simulation....PASS
gc_4x5_CO2_merra2...................................Execute Simulation....PASS
gc_4x5_fullchem_aciduptake_merra2...................Execute Simulation....PASS
gc_4x5_fullchem_APM_merra2..........................Execute Simulation....PASS
gc_4x5_fullchem_benchmark_merra2....................Execute Simulation....PASS
gc_4x5_fullchem_complexSOA_merra2...................Execute Simulation....PASS
gc_4x5_fullchem_complexSOA_SVPOA_merra2.............Execute Simulation....PASS
gc_4x5_fullchem_geosfp..............................Execute Simulation....PASS
gc_4x5_fullchem_marinePOA_merra2....................Execute Simulation....PASS
gc_4x5_fullchem_merra2..............................Execute Simulation....PASS
gc_4x5_fullchem_merra2_47L..........................Execute Simulation....PASS
gc_4x5_fullchem_merra2_LuoWd........................Execute Simulation....PASS
gc_4x5_fullchem_RRTMG_merra2........................Execute Simulation....PASS
gc_4x5_fullchem_TOMAS15_merra2_47L..................Execute Simulation....PASS
gc_4x5_fullchem_TOMAS40_merra2_47L..................Execute Simulation....PASS
gc_4x5_Hg_merra2....................................Execute Simulation....PASS
gc_4x5_metals_merra2................................Execute Simulation....PASS
gc_4x5_POPs_BaP_merra2..............................Execute Simulation....PASS
gc_4x5_tagCH4_merra2................................Execute Simulation....PASS
gc_4x5_tagCO_merra2.................................Execute Simulation....PASS
gc_4x5_tagO3_merra2.................................Execute Simulation....PASS
gc_4x5_TransportTracers_merra2......................Execute Simulation....PASS
gc_4x5_TransportTracers_merra2_LuoWd................Execute Simulation....PASS
 
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 29
Execution tests failed: 0
Execution tests not yet completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

```
==============================================================================
GCHP: Execution Test Results

Number of execution tests: 3
==============================================================================
 
Execution tests:
------------------------------------------------------------------------------
gchp_fullchem_benchmark_merra2_c48..................Execute Simulation....PASS
gchp_fullchem_standard_merra2_c24...................Execute Simulation....PASS
gchp_TransportTracers_geosfp_c24....................Execute Simulation....PASS
 
Summary of execution test results:
------------------------------------------------------------------------------
Execution tests passed:        3
Execution tests failed:        0
Execution tests not completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

File: Interfaces/GCClassic/main.F90 [hunk 1]
```diff
@@ -328,6 +311,9 @@ PROGRAM GEOS_Chem
      CALL Error_Stop( ErrMsg, ThisLoc )
   ENDIF
 
+  ! Set a local variable for Input_Opt%LSETH2O
```
Comment (yantosca) [inline Interfaces/GCClassic/main.F90:314]: We can remove LSETH2O and replace all instances of it with Input_Opt%LSETH2O.

File: run/GCClassic/HISTORY.rc.templates/HISTORY.rc.carbon [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+             #'BoundaryConditions',
+::
+###############################################################################
+### The rest of this file consists of collection definitions.               ###
+### Above collections whose declarations are commented out will be ignored. ###
+### Make sure that each collection definition, as well as the list of       ###
+### collections above, ends with a double-colon ("::").                     ###
+###############################################################################
+
+#==============================================================================
```
Comment (yantosca) [inline run/GCClassic/HISTORY.rc.templates/HISTORY.rc.carbon:48]: Note that the restart collection does not have Chm_KPPHValue, so we don't need to allocate that array.

File: run/GCHP/ExtData.rc.templates/ExtData.rc.carboncycle [hunk 3]
```diff
@@ -0,0 +1,535 @@
+Ext_AllowExtrap: .true.
```
Comment (yantosca) [inline run/GCHP/ExtData.rc.templates/ExtData.rc.carboncycle:1]: This should be renamed to ExtData.rc.carbon.

**Review by yantosca — CHANGES_REQUESTED** [review]
This looks good @msulprizio.,  Just a couple of minor things in the comments to fix and then we're good to merge!

File: run/GCHP/ExtData.rc.templates/ExtData.rc.carboncycle (hunk 3 above)
Comment (msulprizio) [reply run/GCHP/ExtData.rc.templates/ExtData.rc.carboncycle]: Thanks for catching that. This is now fixed in https://github.com/geoschem/geos-chem/pull/1605/commits/ce9b096b1d63735625f6eb1ee459d44494babd11.

File: run/GCClassic/HISTORY.rc.templates/HISTORY.rc.carbon (hunk 2 above)
Comment (msulprizio) [reply run/GCClassic/HISTORY.rc.templates/HISTORY.rc.carbon:48]: Noted. Thanks!

File: Interfaces/GCClassic/main.F90 (hunk 1 above)
Comment (msulprizio) [reply Interfaces/GCClassic/main.F90]: Now fixed in https://github.com/geoschem/geos-chem/pull/1605/commits/ce9b096b1d63735625f6eb1ee459d44494babd11.

File: Headers/state_chm_mod.F90 [hunk 4]
```diff
@@ -1552,13 +1552,12 @@ SUBROUTINE Init_State_Chm( Input_Opt, State_Chm, State_Grid, RC )
     ENDIF ! ITS_A_FULLCHEM_SUM or ITS_AN_AEROSOL_SIM
 
     !========================================================================
-    ! Allocate and initialize fields only needed for FULLCHEM simulations
+    ! Allocate and initialize KPPHvalue (used by KPP-based simulations)
     !========================================================================
-    IF ( Input_Opt%ITS_A_FULLCHEM_SIM .or. Input_Opt%ITS_A_MERCURY_SIM) THEN
+    IF ( Input_Opt%ITS_A_FULLCHEM_SIM      .or.                              &
+         Input_Opt%ITS_A_MERCURY_SIM       .or.                              &
+         Input_Opt%ITS_A_CARBON_SIM      ) THEN
```
[the remark this answers is not in the document]
Comment (msulprizio) [reply Headers/state_chm_mod.F90]: Now fixed in https://github.com/geoschem/geos-chem/pull/1605/commits/ce9b096b1d63735625f6eb1ee459d44494babd11.

**Review by yantosca — APPROVED** [review]
Thanks @msulprizio!  good to go

**Comment by msulprizio** [comment]
This PR has been merged into `dev/14.1.0`. Validation of the carbon simulation against the single-species simulations is ongoing.

**Pull request merged**